File: app_watcher/src/main/java/com/tongchen/twatcher/mzitu/presenter/MZiTuCategoryPresenter.java
package com.tongchen.twatcher.mzitu.presenter;

import android.annotation.SuppressLint;

import com.tongchen.twatcher.base.http.IAppApiHelper;
import com.tongchen.twatcher.base.presenter.MVPPresenter;
import com.tongchen.twatcher.mzitu.model.entity.MZiTu;
import com.tongchen.twatcher.mzitu.view.IMZiTuCategoryView;
import com.tongchen.twatcher.util.LogUtils;

import java.util.List;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

/**
 * Created by TongChen at 11:34 on 2018/12/28.
 * <p>
 * Description: 妹子图按分类获取封面列表
 */
public class MZiTuCategoryPresenter extends MVPPresenter<IMZiTuCategoryView, List<MZiTu>> implements IMZiTuCategoryPresenter {

    private IAppApiHelper mIAppApiHelper;
    //  请求的方式：0 refresh;1 more
    private int mMode = -1;
    public static final int MODE_REFRESH = 0;
    public static final int MODE_MORE = 1;

    private int page = 1;
    private int totalPage = 1;

    public MZiTuCategoryPresenter(IAppApiHelper iAppApiHelper) {
        mIAppApiHelper = iAppApiHelper;
    }

    @SuppressLint("CheckResult")
    @Override
    public void getMZiTuCoverListByCategory(String category, boolean pullToRefresh) {
        if (pullToRefresh) {
            page = 1;
        }
        mIAppApiHelper.listMZiTu(category, page, pullToRefresh)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(listBaseResult -> {
                            requestSucceed(listBaseResult.getData());
                        },
                        throwable -> requestFailed(throwable.toString()));
    }

    @Override
    public void requestSucceed(List<MZiTu> result) {
        if (getView() == null) {
            return;
        }
        LogUtils.d("MZiTuCategoryPresenter", "requestSucceed:" + result.toString());
        mMode=0;
        switch (mMode) {
            case MODE_REFRESH:
                getView().hideLoading();
                getView().refreshSucceed(result);
                break;
            case MODE_MORE:
                getView().loadMoreSucceed(result);
                break;
            default:
                break;
        }
    }

    @Override
    public void requestFailed(String errorMsg) {
        if (getView() == null) {
            return;
        }
        LogUtils.d("MZiTuCategoryPresenter", "requestFailed:" + errorMsg);

        switch (mMode) {
            case MODE_REFRESH:
                getView().hideLoading();
                getView().refreshFailed(errorMsg);
                break;
            case MODE_MORE:
                getView().loadMoreFailed(errorMsg);
                break;
            default:
                break;
        }
    }
}

File: app_watcher/src/main/java/com/tongchen/twatcher/mzitu/presenter/MZiTuCategoryPresenter.kt
package com.tongchen.twatcher.mzitu.presenter

import android.annotation.SuppressLint
import com.tongchen.twatcher.base.http.IAppApiHelper
import com.tongchen.twatcher.base.presenter.MVPPresenter
import com.tongchen.twatcher.gank.view.ICategoryView
import com.tongchen.twatcher.mzitu.model.entity.MZiTu
import com.tongchen.twatcher.mzitu.view.IMZiTuCategoryView
import com.tongchen.twatcher.util.LogUtils
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers

/**
 * @author TongChen
 * @date 2019/11/28  14:55
 * <p>
 * Desc: 妹子图按分类获取封面列表
 */
class MZiTuCategoryPresenter : MVPPresenter<IMZiTuCategoryView, MutableList<MZiTu>>, IMZiTuCategoryPresenter {

    private val TAG = "MZiTuCategoryPresenter"

    private var mIAppApiHelper: IAppApiHelper
    //  请求的方式：0 refresh;1 more
    private var mMode = -1
    val MODE_REFRESH = 0
    val MODE_MORE = 1

    private var mPage = 1
    private var mTotalPage = 1

    constructor(mIAppApiHelper: IAppApiHelper) : super() {
        this.mIAppApiHelper = mIAppApiHelper
    }

    @SuppressLint("CheckResult")
    override fun getMZiTuCoverListByCategory(category: String?, pullToRefresh: Boolean) {
        if (pullToRefresh) {
            mMode = MODE_REFRESH
            mPage = 1

        } else {
            mMode = MODE_MORE
            if (mPage < mTotalPage) {
                mPage++
            }
        }

        mIAppApiHelper.listMZiTu(category, mPage, pullToRefresh)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ baseResult ->
                    if (baseResult != null) {
                        mTotalPage = baseResult.totalPage
                        requestSucceed(baseResult.data)
                    } else {
                        requestFailed("data is wrong")
                    }
                }, { throwable -> requestFailed(throwable.toString()) })
    }

    override fun requestSucceed(result: MutableList<MZiTu>?) {
        if (view == null) {
            return
        }
        LogUtils.d(TAG, "requestSucceed:${result.toString()}")

        when (mMode) {
            MODE_REFRESH -> {
                (view as IMZiTuCategoryView).hideLoading()
                (view as IMZiTuCategoryView).refreshSucceed(result)
            }

            MODE_MORE -> {
                (view as IMZiTuCategoryView).loadMoreSucceed(result)
            }
        }

    }

    override fun requestFailed(errorMsg: String) {
        if (view == null) {
            return
        }
        LogUtils.d(TAG, "requestFailed:$errorMsg")

        when (mMode) {
            MODE_REFRESH -> {
                (view as IMZiTuCategoryView).hideLoading()
                (view as IMZiTuCategoryView).refreshFailed(errorMsg)
            }

            MODE_MORE->{
                (view as IMZiTuCategoryView).loadMoreFailed(errorMsg)
            }
        }
    }


}